### PR TITLE
Support more authentication algorithms in OSPF dissector

### PR DESCRIPTION
**Pull request by kholia**
https://c0decafe.de/svn/codename_loki/test/ has sample pcap files for testing.

The existing code only supports MD5 type hashes. This commit adds support for HMAC-SHA1, HMAC-SHA256, HMAC-SHA384 and HMAC-SHA512 hashes.

Here is the patch in action,

```
$ ./src/ettercap -Tqr ospf-hmac-sha-256_key-1234.pcap 
...
OSPF-224.0.0.5-0:$ospf$2$0201002cac10000a00000000<...>$508a1abffb5b4554e1aa46eb05<...>
```

Such output hashes can be cracked using latest JtR jumbo. For example,

```
$ ettercap -Tqr ospf-hmac-sha-256_key-1234.pcap > hash

$ john --format=ospf hash 
Using default input encoding: UTF-8
Loaded 1 password hash (ospf, OSPF [HMAC-SHA-X 32/64])
Will run 4 OpenMP threads
Press 'q' or Ctrl-C to abort, almost any other key for status
1234             (OSPF-224.0.0.5-0)
...
Session completed
```

### CODE REVIEW

File: src/dissectors/ec_ospf.c [hunk 1]
```diff
@@ -140,9 +152,24 @@ FUNC_DECODER(dissector_ospf)
...
+        } else if (auth_data_len == OSPF_AUTH_HMAC_SHA384_SIZE) {
+                type = 3;
+        } else if (auth_data_len == OSPF_AUTH_HMAC_SHA512_SIZE) {
+                type = 4;
+        }
+        if (type != 0) {
+                DISSECT_MSG("OSPF-%s-%d:$ospf$%d$",
+                        ip_addr_ntoa(&PACKET->L3.dst, tmp),
+                        ntohs(PACKET->L4.dst), type);
+        }
```
Comment (koeppea) [inline src/dissectors/ec_ospf.c]: Wouldn't this be easier in a *switch/case* statement?
```c
switch (auth_data_len) {
   case OSPF_AUTH_MD5_SIZE:
      DISSECT_MSG(...);
      break;
   case OSPF_AUTH_HMAC_SHA1_SIZE:
   case OSPF_AUTH_HMAC_SHA256_SIZE:
   case OSPF_AUTH_HMAC_SHA512_SIZE:
      DISSECT_MSG(...);
   break;
}
```
A *default*-block with a `return NULL` may even allow to do the check of the first *if*-statement all in one.
This way you also don't require the **type** variable.

Comment (kholia) [inline src/dissectors/ec_ospf.c]: The switch construct looks nicer. I will switch to using it soon. Thanks!

Comment (kholia) [inline src/dissectors/ec_ospf.c]: @koeppea I require the `type` variable or at least I require its value. It is part of the output hash string. I can't see an easy way to use the switch construct and still keep the `type` logic intact.

Comment (kholia) [inline src/dissectors/ec_ospf.c]: I have removed the following code in the latest commit,

```
if (auth_data_len != OSPF_AUTH_MD5_SIZE
    && auth_data_len != OSPF_AUTH_HMAC_SHA1_SIZE
    && auth_data_len != OSPF_AUTH_HMAC_SHA256_SIZE
    && auth_data_len != OSPF_AUTH_HMAC_SHA384_SIZE
    && auth_data_len != OSPF_AUTH_HMAC_SHA512_SIZE) {
        return NULL;
}
```

This code piece was ugly and is not required.

Comment (koeppea) [inline src/dissectors/ec_ospf.c]: > @koeppea I require the type variable or at least I require its value. It is part of the output hash string. I can't see an easy way to use the switch construct and still keep the type logic intact.

I see. So the type value is required to identify the hash type for *john* or *jack*?
I agree, that with maintaining this **type** value, a switch/case doesn't make much sense then.

Comment (kholia) [inline src/dissectors/ec_ospf.c]: Yes, john (JtR) uses this `type` value to automatically identify the hash type.

https://github.com/magnumripper/JohnTheRipper/blob/bleeding-jumbo/src/ospf_fmt_plug.c

Thanks! :)

**Comment by koeppea** [comment]
Merged. Thanks.

**Pull request merged**